pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// filterFrom returns a filtered list of instance types based on the provided resource constraints
+func (p *InstanceTypeProvider) filterFrom(instanceTypes []*ec2.InstanceTypeInfo, constraints *cloudprovider.Constraints) []*ec2.InstanceTypeInfo {
+	filteredInstancePools := []*ec2.InstanceTypeInfo{}
+	architecture := "x86_64"
+	if constraints.Architecture != nil && *constraints.Architecture == v1alpha1.ArchitectureArm64 {
+		architecture = string(*constraints.Architecture)
+	}
+
+	for _, instanceTypeInfo := range instanceTypes {
+		if len(constraints.InstanceTypes) != 0 {
+			if !functional.ContainsString(constraints.InstanceTypes, *instanceTypeInfo.InstanceType) {
+				continue
+			}
+		}
+		if !functional.ContainsString(aws.StringValueSlice(instanceTypeInfo.ProcessorInfo.SupportedArchitectures), architecture) {
+			continue
+		}
+		if !functional.ContainsString(aws.StringValueSlice(instanceTypeInfo.SupportedUsageClasses), "on-demand") {
+			continue
+		}
+		if *instanceTypeInfo.BurstablePerformanceSupported {
+			continue
+		}
+		if instanceTypeInfo.FpgaInfo != nil {
+			continue
+		}
+		if instanceTypeInfo.GpuInfo != nil {
+			continue
+		}
+		filteredInstancePools = append(filteredInstancePools, instanceTypeInfo)
+	}
+	return filteredInstancePools
+}
+
+// filterByLocation returns a list of instance types that are supported in the provided availability zone using the ec2 DescribeInstanceTypeOfferings API
+func (p *InstanceTypeProvider) filterByLocation(ctx context.Context, instanceTypes []*ec2.InstanceTypeInfo, zone string) ([]*ec2.InstanceTypeInfo, error) {
+	inputs := &ec2.DescribeInstanceTypeOfferingsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("location"),
+				Values: aws.StringSlice([]string{zone}),
+			},
+		},
+		LocationType: aws.String("availability-zone"),
+	}
+
+	instanceTypeNamesSupported := []string{}
+
+	err := p.ec2.DescribeInstanceTypeOfferingsPagesWithContext(ctx, inputs, func(output *ec2.DescribeInstanceTypeOfferingsOutput, lastPage bool) bool {
+		for _, offerings := range output.InstanceTypeOfferings {
+			instanceTypeNamesSupported = append(instanceTypeNamesSupported, *offerings.InstanceType)
+		}
+		return true
+	})
+	if err != nil {
+		return instanceTypes, fmt.Errorf("describing instance type location offerings, %w", err)
+	}
+
+	instanceTypeInfoSupported := []*ec2.InstanceTypeInfo{}
+	for _, instanceTypeInfo := range instanceTypes {
+		for _, instanceTypeName := range instanceTypeNamesSupported {
+			if instanceTypeName == *instanceTypeInfo.InstanceType {
+				instanceTypeInfoSupported = append(instanceTypeInfoSupported, instanceTypeInfo)
+			}
+		}
+	}
+	return instanceTypeInfoSupported, nil
+}
+
+// normalizeZones takes zone names or ids and returns them all as zone names
+func (p *InstanceTypeProvider) normalizeZones(ctx context.Context, zones []string) ([]string, error) {
+	azs, err := p.ec2.DescribeAvailabilityZonesWithContext(ctx, &ec2.DescribeAvailabilityZonesInput{})
```

Comment:
Should we cache zone names to avoid making this call frequently?

Reply:
yes, that's a good call! And maybe put this in the vpc provider too...